src/xcode/ENA/ENA/Source/Services/HealthCertificateValidationService/HealthCertificateValidationService.swift

```diff
@@ -172,4 +212,557 @@ final class HealthCertificateValidationService: HealthCertificateValidationProvi
+			completion(.failure(.ACCEPTANCE_RULE_SERVER_ERROR))
+		}
+	}
+	
+	private func acceptanceRules(_ data: Data, completion: (Result<[Rule], HealthCertificateValidationError>) -> Void) {
+		let extractAcceptanceRulesResult = ValidationRulesAccess().extractValidationRules(from: data)
+		
+		switch extractAcceptanceRulesResult {
+		case let .success(acceptanceRules):
+			completion(.success(acceptanceRules))
+		case .failure:
+			completion(.failure(.ACCEPTANCE_RULE_JSON_DECODING_FAILED))
+		}
+	}
+	
+	// MARK: - Invalidation Rules
+	
+	private func downloadInvalidationRule(
+		completion: @escaping (Result<[Rule], HealthCertificateValidationError>) -> Void
+	) {
+		client.getDCCRules(
+			eTag: store.invalidationRulesCache?.lastValidationRulesETag,
+			isFake: false,
+			ruleType: .invalidation,
+			completion: { [weak self] result in
+				guard let self = self else {
+					Log.error("Could not create strong self", log: .vaccination)
+					completion(.failure(.INVALIDATION_RULE_CLIENT_ERROR))
+					return
+				}
+				
+				switch result {
+				case let .success(packageDownloadResponse):
+					self.invalidationRuleDownloadingSuccessHandler(
+						packageDownloadResponse: packageDownloadResponse,
+						completion: completion
+					)
+	
+				case let .failure(error):
+					self.invalidationRuleDownloadingFailureHandler(
+						error: error,
+						completion: completion
+					)
+				}
+			}
+		)
+	}
+	
+	private func invalidationRuleDownloadingSuccessHandler(
+		packageDownloadResponse: PackageDownloadResponse,
+		completion: @escaping (Result<[Rule], HealthCertificateValidationError>) -> Void
+	) {
+		Log.info("Successfully received invalidation rules package. Proceed with eTag verification...", log: .vaccination)
+
+		guard let eTag = packageDownloadResponse.etag else {
+			Log.error("ETag of package is missing. Return with failure.", log: .vaccination)
+			completion(.failure(.INVALIDATION_RULE_JSON_ARCHIVE_SIGNATURE_INVALID))
+			return
+		}
+		
+		Log.info("Successfully verified eTag. Proceed with package extraction...", log: .vaccination)
+				
+		guard !packageDownloadResponse.isEmpty,
+			  let sapDownloadedPackage = packageDownloadResponse.package else {
+			Log.error("PackageDownloadResponse is empty. Return with failure.", log: .vaccination)
+			completion(.failure(.INVALIDATION_RULE_JSON_ARCHIVE_FILE_MISSING))
+			return
+		}
+		Log.info("Successfully extracted sapDownloadedPackage. Proceed with package verification...", log: .vaccination)
+		
+		guard self.signatureVerifier.verify(sapDownloadedPackage) else {
+			Log.error("Verification of sapDownloadedPackage failed. Return with failure", log: .vaccination)
+			completion(.failure(.INVALIDATION_RULE_JSON_ARCHIVE_SIGNATURE_INVALID))
+			return
+		}
+		Log.info("Successfully verified sapDownloadedPackage. Proceed now with CBOR decoding...", log: .vaccination)
+		
+		
+		self.acceptanceRules(sapDownloadedPackage.bin, completion: { result in
```

Comment:
```suggestion
		extractInvalidationRules(sapDownloadedPackage.bin, completion: { result in
```

Called acceptance rules here, but I guess all it does is lead to the wrong error code

Reply:
Correct, wrong call. I hope I would have found it when writing the unit tests.